hack/verify-e2e-test-all-presence.sh

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
+
+# This script verifies that all the e2e tests defined in package
+# test/e2e have a corresponding invocation in the TestAll function
+# defined in test/e2e/all_test.go.
+#
+# The TestAll function provides ordering of all e2e tests. Tests that
+# can be run in parallel are invoked first and will run to completion
+# before starting those that must run serially.
+#
+# The CI job runs `make verify` before starting the e2e tests and the
+# verify target will run this script. If this script detects any Go
+# Test function by name that is not invoked by the TestAll function
+# then it will list the omission and exit with an error, preventing
+# the e2e tests from starting.
+
+# This script has been tested on Linux and macOS.
+
+set -u
+set -o pipefail
+
+thisdir="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd -P)"
+e2e_test_dir="$thisdir/../test/e2e"
+e2e_test_file="$e2e_test_dir/all_test.go"
+
+if ! [[ -f ${e2e_test_file} ]]; then
+    echo "error: $e2e_test_file is missing."
+    exit 1
+fi
+
+# "go test -list" must run in the directory where the test files are.
+pushd "$e2e_test_dir" >/dev/null || {
+    echo "error: pushd $e2e_test_dir failed" >&2;
+    exit 2                      # ENOENT
+}
+
+go_test_list_output=$(E2E_TEST_MAIN_SKIP_SETUP=1 go test -list . -tags e2e)
```

Comment:
Would it make sense to add a "list-e2e-tests" target?  Doing so would have two advantages:

* The user could use `make list-e2e-tests` to get a list of tests.
* Someone modifying the `test-e2e` target in `Makefile` would be more likely to notice the `list-e2e-tests` target and update it as well as needed.

Reply:
More than the following? 

```Makefile
.PHONY: test-e2e-list
test-e2e-list:
	@(cd ./test/e2e; E2E_TEST_MAIN_SKIP_SETUP=1 $(GO) test -list . -tags e2e | grep ^Test)
```

```console
$ /bin/bash -c 'time make test-e2e-list'                    
TestAll
TestCanaryRoute
TestCreateIngressControllerThenSecret
TestCreateSecretThenIngressController
TestHealthCheckIntervalIngressController
TestClientTLS
TestConfigurableRouteRBAC
TestConfigurableRouteNoSecretNoRBAC
TestConfigurableRouteNoConsumingUserNoRBAC
TestIngressStatus
TestForwardedHeaderPolicyAppend
TestForwardedHeaderPolicyReplace
TestForwardedHeaderPolicyNever
TestForwardedHeaderPolicyIfNone
TestHAProxyTimeouts
TestHAProxyTimeoutsRejection
TestRouteHardStopAfterEnableOnIngressConfig
TestRouteHardStopAfterEnableOnIngressController
TestRouteHardStopAfterEnableOnIngressControllerHasPriorityOverIngressConfig
TestRouteHardStopAfterTestInvalidDuration
TestRouteHardStopAfterTestZeroLengthDuration
TestRouteHardStopAfterTestOneDayDuration
TestHstsPolicyWorks
TestRouteHTTP2EnableAndDisableIngressController
TestRouteHTTP2EnableAndDisableIngressConfig
TestHTTPHeaderBufferSize
TestHeaderNameCaseAdjustment
TestTunableMaxConnectionsValidValues
TestTunableMaxConnectionsInvalidValues
TestRouteNbthreadIngressController
TestOperatorSteadyConditions
TestClusterOperatorStatusRelatedObjects
TestDefaultIngressControllerSteadyConditions
TestDefaultIngressClass
TestCustomIngressClass
TestUserDefinedIngressController
TestUniqueDomainRejection
TestProxyProtocolOnAWS
TestProxyProtocolAPI
TestUpdateDefaultIngressController
TestIngressControllerScale
TestDefaultIngressCertificate
TestPodDisruptionBudgetExists
TestHostNetworkEndpointPublishingStrategy
TestHostNetworkPortBinding
TestInternalLoadBalancer
TestInternalLoadBalancerGlobalAccessGCP
TestScopeChange
TestNodePortServiceEndpointPublishingStrategy
TestTLSSecurityProfile
TestRouteAdmissionPolicy
TestSyslogLogging
TestContainerLogging
TestIngressControllerCustomEndpoints
TestHTTPHeaderCapture
TestHTTPCookieCapture
TestNetworkLoadBalancer
TestUniqueIdHeader
TestLoadBalancingAlgorithmUnsupportedConfigOverride
TestDynamicConfigManagerUnsupportedConfigOverride
TestLocalWithFallbackOverrideForLoadBalancerService
TestLocalWithFallbackOverrideForNodePortService
TestReloadIntervalUnsupportedConfigOverride
TestCustomErrorpages
TestTunableRouterKubeletProbesForCustomIngressController
TestIngressControllerServiceNameCollision
TestRouterCompressionParsing
TestRouterCompressionOperation

real	0m1.594s
user	0m2.014s
sys	0m0.731s
```

Reply:
Added in 5e93c23a92d68bc645da29518b3511a0443c24bc.

Reply:
The point was that you could use `make test-e2e-list` in this shell script.  Any reason not to do that?